src/style-spec/style-spec.js

```diff
@@ -1,51 +1,46 @@
 // @flow
 
+type ExpressionType = 'data-driven' | 'cross-faded' | 'cross-faded-data-driven' | 'color-ramp' | 'data-constant';
+type ExpressionParameters = Array<'zoom' | 'feature' | 'heatmap-density' | 'line-progress'>;
+
+type ExpressionSpecification = {
+    type: ExpressionType,
```

Comment:
Looks like field is currently called `property-type`, not `type`

Reply:
It feels slightly weird to me to have `property-type` nested under `expression`.  Would it make sense to move it up to the layout/paint property specification, and also add `constant` as one of the options for properties that can't take any expression?   E.g.:

```js
"visibility": {
  "property-type": "constant",
  // no "expression" field
  "type": "enum",
   /* ... */
}
```

```js
"line-cap": {
  "property-type": "data-constant",
  "expression": {
    "interpolated": false,
    "parameters": ["zoom"]
   },
  "type": "enum",
   /* ... */
}
```

Then we could always use `propertySpec["property-type"]` and never have to rely on the presence/absence of the "expression" field

Reply:
Sounds good to me @anandthakker!